### ISSUE TITLE
How to get good results on PTB dataset?

### ISSUE THREAD

**Issue by rivaldinho** [tensorflow/tensor2tensor#128]
similar results like #108 when running with
PROBLEM=lmptb_10k
MODEL=attention_lm
HPARAMS=attention_lm_base

> Still got similar results, so I tried to turn down the hidden_size and num_hidden_layers, also increased the attention_dropout and residual_dropout, but still no satisfactory results with log_perplexity around 5.00.

What I got is either overfitting or log_perplexity about 5.1 (training loss is jitter around 4.3), which makes me really confused. I noticed there are some specific configurations in Table 3 of the paper "Attention is all you need", so can you list the proper configurations to achieve better results on PTB？Which parameters need to be tuned other than dropout or hidden_size?
Please help.

**Comment by lukaszkaiser** [comment]
You definitely need to lower filter_size and hidden_size and set all dropouts higher. Also, as far as I can tell, log-ppx of 4.4 is quite good and 4.1 would be state-of-the-art, so 5.0 isn't that bad for a model that's totally untuned for the task. Note that we never ran our models on PTB as it's a very small data-set and it was only added recently in a github PR. I'm just giving it a shot with these params:

--problems=lmptb_10k --model=attention_lm --hparams_set=attention_lm_base --train_steps=20000 --eval_steps=7 --hparams='hidden_size=512,filter_size=1024,dropout=0.5,attention_dropout=0.5,label_smoothing=0.1,relu_dropout=0.5,residual_dropout=0.5,learning_rate_warmup_steps=2000,learning_rate_decay_scheme=exp10k'

I'm getting 4.8 after 8k steps, so not too horrible, but still way to go!

If you get a better hparams set for ptb, please make a PR adding it to attention_lm as attention_lm_ptb :). It's also sth to collaborate on with others! Please let me know what results you see.

**Comment by rivaldinho** [comment]
@lukaszkaiser 
Thanks! I'll try to find a better hparams set for PTB.
I'm also trying to reproduce good results for the 1B words benchmark, and here are 2 questions I have.
1) What's the best result you've got on the 1B words benchmark and what's the specific params for this best result?

2) What's the vocabulary size used in your experiment? If it's 800k, have you applied some methods to speedup the softmax computation for large vocabulary size like Hierarchical Softmax or some other sampling methods? Since as far as I know, it would take a long time to train this model without such acceleration.

**Comment by lukaszkaiser** [comment]
Hi @rivaldinho : it'll be great if you find some good params for PTB, let us know.

As for LM1B: Noam just added in 1.0.13 a generator for the lm1b dataset with 32k subword vocabulary. I think that's the setting he got the best results with, I think attention_lm_moe goes below 30 ppx. I can ask him for a detailed number, but of course these close-to-SOTA LMs take a while to train even on 8 gpus.

**Comment by fminkin** [comment]
Hello @lukaszkaiser 
Can you explain the purpose of that 32k subword vocabulary? How is it used?

**Comment by martinpopel** [comment]
@fminkin: Subword units (either [BPE](https://arxiv.org/abs/1508.07909) or [word-pieces](https://arxiv.org/abs/1609.08144), which is almost the same) are a compromise between character-based models (which has small "vocabulary", but long sequences) and word-based models (English may have more than 200k wordforms, which would result in big models and slow training).

**Comment by fminkin** [comment]
@martinpop, thanks for clearing that out. 
So it's beneficial to use subwords in combination with words as word level language model? How exactly does it work?

**Comment by lukaszkaiser** [comment]
Subword vocabulary contains the most common words anyway, so there is no need to merge. In general, subwords are probably the best option for text problems. But PTB is an old dataset and it comes pre-tokenized to a fixed vocabulary anyway, so it's not a good example for that.

**Comment by sainivarsha97** [comment]
my google colab is showing runtime died  for large vocab size
